Show resolver being used when `stack ghci` is invoked outside of a project directory

wafelj opened the pull request:
Fix for https://github.com/commercialhaskell/stack/issues/3651.

Approach: read `Project` from `Config`, get `SnapshotLocation` and compiler info from it and use that to get resolver using the `loadResolver` function. To my understanding it is theoretically possible that there is no project information in config (it's wrapped in `Maybe`). But I'm assuming this never actually happens because if I try to delete the `~/.stack/global-project/stack.yaml`, a new one is recreated when `stack ghci` is run again. This is why I assumed it's safe to use `fromJust`. If there is a better approach, I'd love to hear about it.

Disclaimer: I'm new to Haskell and to stack, and this is my first open-source contribution. Any feedback is most appreciated!

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

mgsloan commented on src/Stack/Ghci.hs:
```diff
@@ -186,8 +189,12 @@ ghci opts@GhciOpts{..} = do
+
     buildDepsAndInitialSteps opts (map (T.pack . packageNameString . fst) localTargets)
-    targetWarnings stackYaml localTargets nonLocalTargets mfileTargets
+
```
Please don't insert newlines here unless it's logically splitting things into sections (also above)

wafelj replied:
Removed.

mgsloan commented on src/Stack/Ghci.hs:
```diff
@@ -836,13 +843,14 @@ checkForDuplicateModules pkgs = do
+  -> Maybe (Project, Path Abs File)
   -> RIO env ()
-targetWarnings stackYaml localTargets nonLocalTargets mfileTargets = do
+targetWarnings stackYaml localTargets nonLocalTargets mfileTargets mproject = do
```
Since the context has been changed from `HasRunner` to `HasEnvConfig`, there is no need to pass in `mproject`, instead `view $ configL.to configMaybeProject` can be done in this function.  Similarly, no need to pass in `stackYaml`, that can also be queried from the reader environment.

wafelj replied:
Makes sense, done.

mgsloan commented:
The code related to resolvers and snapshots has changed quite a bit since I last looked, so I can't give direct guidance.  However, it seems like `loadResolver` does fetching and parsing that I believe should have already happened.  So, I think it'd be nice to avoid the redundant call, or cache it.  I'll leave that decision up to the more active maintainers, though.

wafelj commented on src/Stack/Ghci.hs:
```diff
@@ -853,10 +861,16 @@ targetWarnings stackYaml localTargets nonLocalTargets mfileTargets = do
-  when (null localTargets && isNothing mfileTargets) $
+
+  when (null localTargets && isNothing mfileTargets) $ do
+      let project = fst $ fromJust mproject
```
I agree, ignoring the `Nothing` case is pretty ugly. I was thinking about this some more and decided to try something else: simply not print the resolver version if the project is `Nothing`. Would this work? I can do the `error` instead, but it's basically the same thing I had before, except it has a more useful error message.

wafelj commented:
@mgsloan thanks for your review! I was digging through the `HasEnvConfig` environment, trying to see if and where `SnapshotDefinition` is available, but I could not find it. `SnapshotLocation` inside `Project` inside `Config` was the closest, so I went with that. Definitely interested in other suggestions - my understanding of the project is not very deep yet.

mgsloan commented:
@wafelj Welcome!  Yeah I also dug around `HasEnvConfig` looking for snapshot name info.  Couldn't seem to find it, which seems surprising since `EnvConfig` has `BuildConfig`, and info about the snapshot needs to be loaded to have a `BuildConfig`.  So, yeah, not sure why that's not there ¯\_(ツ)_/¯

Looks to me like `bcSnapshot :: LoadedSnapshot` could simply be added to `BuildConfig` and then populated in `loadbuildConfig`.

qrilka commented:
@mgsloan @wafelj there was a large refactoring in #4412 and now Stack doesn' use `LoadedSnapshot` anymore. I think snapshot name could be added into `SMWanted` which later gets transformed into `SourceMap` - snapshot gets loaded in `Stack.Config.loadBuildConfig` but its name gets ignored (only compiler/dependencies get used).

wafelj commented:
@mgsloan @qrilka thanks. I looked into that and changed the code - now it passes the name from `SMWanted` to `SMActual` to `SourceMap`. Then I load `SourceMap` from `EnvConfig` and read the name from that. Is this what you had in mind?

Also, two questions:
1) `SMWanted` is available in in `BuildConfig`. Would it make sense to read it directly from that in `targetWarnings` function instead of passing it through all those other types?
2) naming - is `smwName` etc. descriptive enough or would `smwResolverName` or `smwSnapshotName` etc. be better?

qrilka commented:
1. The name should be the same in either of `SMXXX` or `SourceMap` so using `BuildConfig` should be OK
2. I think we should use longer field name as it's not a name of a source map itself. Also it makes sense to follow https://github.com/commercialhaskell/stack/issues/4256 and use "snapshot".
@snoyberg should we also switch to use "snapshot" instead of "resolver" in Stack output?

snoyberg commented:
> should we also switch to use "snapshot" instead of "resolver" in Stack output?

Sounds like a good idea to me.

wafelj commented:
@qrilka changed to read name directly from BuildConfig and changed the field to smwSnapshotName. Also changed output to snapshot instead of resolver. Looks good?
